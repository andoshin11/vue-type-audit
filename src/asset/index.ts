export const GLOBAL_TYPES_FILE = {
  name: '^__GLOBAL_TYPES.d.ts',
  identifier: '^__GLOBAL_TYPES',
  content: `
import { ComponentPublicInstance, VNodeProps, VNode, VNodeTypes, ComponentOptions, DefineComponent, Prop } from 'vue'

export type ClassInstance<T> = T extends new (...args: any[]) => infer U ? U : never;

export type isNeverType<T> = [T] extends [never] ? true : false

type InferPropType<T> = T extends null ? any : T extends {
  type: null | true;
} ? any : T extends ObjectConstructor | {
  type: ObjectConstructor;
} ? Record<string, any> : T extends BooleanConstructor | {
  type: BooleanConstructor;
} ? boolean : T extends Prop<infer V, infer D> ? (unknown extends V ? D : V) : T;


type __ExtractPropTypes<O> = O extends object ? {
  [K in _RequiredProps<O>]: InferPropType<O[K]>;
} & {
  // @ts-ignore
  [K in Exclude<keyof O, _RequiredProps<O>>]?: InferPropType<O[K]>;
} : {
  [K in string]: any;
};

export type PropTypes<C> = C extends DefineComponent<infer PropsOptions, infer RawBindings, infer D, any, infer M, infer Mixin, infer Extends, infer E, infer EE> ? __ExtractPropTypes<PropsOptions> : never

export type EmitTypes<C> = C extends DefineComponent<infer P, infer B, infer D, any, infer M, infer Mixin, infer Extends, infer E, infer EE, infer PublicProps, infer Props, infer Defaults> ? E : never

export type ElementType<T> = T extends (infer R)[] ? R : never

export type ToHandlerType<F> = F extends (...args: infer A) => boolean ? (...args: A) => void : null extends F ? (...args: any) => void : never;

export type WithEmitType<T, D extends Record<string, (keyof EmitTypes<T>) | (ElementType<EmitTypes<T>>)>> = {
  __emitHandlerTypes: {
    [K in keyof D]?: false extends isNeverType<ElementType<EmitTypes<T>>> ? (...args: any) => any : D[K] extends keyof EmitTypes<T> ? ToHandlerType<EmitTypes<T>[D[K]]> : never
  }
}

export type MixIntoComponent<C, T> = C extends DefineComponent<infer P, infer B, infer D, any, infer M, infer E, infer PublicProps, infer Defaults, infer MakeDefaultsOptional, infer Options> ? DefineComponent<P, B, D, any, M, E, PublicProps, Defaults, MakeDefaultsOptional, Options & T> : never;

type _RequiredProps<T> = {
  [K in keyof T]: T[K] extends {
      required: true;
  } ? K : never;
}[keyof T];

export type RequiredProps<C> = C extends DefineComponent<infer PropsOptions, infer RawBindings, infer D, any, infer M, infer Mixin, infer Extends, infer E, infer EE> ? _RequiredProps<PropsOptions> : never

export declare function _resolveComponent<N>(name: N): N;

export type _VNodeProps = VNodeProps
export type _VNode = VNode
export type _VNodeTypes = VNodeTypes
export interface _ClassComponent {
  new (...args: any[]): ComponentPublicInstance<any, any, any, any, any>;
  __vccOpts: ComponentOptions;
}
export type _Data = {
  [key: string]: unknown;
};
`
}
