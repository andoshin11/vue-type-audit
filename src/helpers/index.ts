export * from './diagnostics'
export * from './node'
export * from './path'
