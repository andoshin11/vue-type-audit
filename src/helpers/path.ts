import {
  RawVueFileName,
  VueTemplateFileName,
  CompiledVueTemplateFileName,
  TransformedCompiledVueTemplateFileName,
  TSVueFileName,
  VueScriptFileName,
  TransformedVueScriptFileName
} from "../types";

export const toRelativePath = (str: string) =>
  str.replace(process.cwd() + "/", "");

export const toTSVueFIleName = (fileName: RawVueFileName) => `${fileName}.ts` as TSVueFileName

export const toVueTemplateFileName = (fileName: RawVueFileName) =>
  `${fileName}.template.ts` as VueTemplateFileName;

export const toVueScriptFileName = (fileName: RawVueFileName) =>
  `${fileName}.script.ts` as VueScriptFileName;

export const toCompiledVueTemplateFileName = (fileName: VueTemplateFileName) =>
  fileName.replace(/\.ts$/, ".compiled.ts") as CompiledVueTemplateFileName;

export const toTransformedCompiledVueTemplateFileName = (
  fileName: CompiledVueTemplateFileName
) =>
  fileName.replace(
    /\.compiled\.ts$/,
    ".compiled.transformed.ts"
  ) as TransformedCompiledVueTemplateFileName;

export const toTransformedVueScriptFileName = (fileName: VueScriptFileName) => fileName.replace(/\.ts$/, '.transformed.ts') as TransformedVueScriptFileName

export const toRawVueFileName = (fileName: TSVueFileName | CompiledVueTemplateFileName): RawVueFileName => {
  if (isTSVueFile(fileName)) {
    return fileName.replace(/\.ts$/, '') as RawVueFileName
  } else if (isCompiledVueTemplateFile(fileName)) {
    return fileName.replace(/\.template\.compiled\.ts$/, '') as RawVueFileName
  } else {
    throw new Error('could not trnsform to raw vue file name')
  }
}

/**
 * User Type Guards
 */
export const isRawVueFile = (fileName: string): fileName is RawVueFileName =>
  /\.vue$/.test(fileName);

export const isTSVueFile = (fileName: string): fileName is TSVueFileName =>
  /\.vue\.ts$/.test(fileName);

export const isVueScriptFile = (fileName: string): fileName is VueScriptFileName => /\.vue\.script\.ts$/.test(fileName)

export const isVueTemplateFile = (fileName: string): fileName is VueTemplateFileName => /\.vue\.template\.ts$/.test(fileName)

export const isCompiledVueTemplateFile = (fileName: string): fileName is CompiledVueTemplateFileName => /\.vue\.template\.compiled\.ts$/.test(fileName)
