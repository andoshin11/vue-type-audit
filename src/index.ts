export * from './asset'
export * from './helpers'
export * from './sourcemap'
export * from './transformers'
export * from './types'
