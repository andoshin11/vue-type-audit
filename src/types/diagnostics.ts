import _ts from 'typescript'

export type DiagnosticWithRange = Omit<_ts.Diagnostic, 'start' | 'length'> & { start: NonNullable<_ts.Diagnostic['start']>; length: NonNullable<_ts.Diagnostic['length']> }
