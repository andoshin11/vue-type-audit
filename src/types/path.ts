import { Brand } from '.'

export type RawVueFileName = Brand<string, 'RawVueFileName'>

export type TSVueFileName = Brand<string, 'TSVueFileName'>

export type VueTemplateFileName = Brand<string, 'VueTemplateFileName'>

export type VueScriptFileName = Brand<string, 'VueScriptFileName'>

export type CompiledVueTemplateFileName = Brand<string, 'CompiledVueTemplateFileName'>

export type CompiledVueScriptFileName = Brand<string, 'CompiledVueTemplateFileName'>

export type TransformedCompiledVueTemplateFileName = Brand<string, 'TransformedCompiledVueTemplateFileName'>

export type TransformedVueScriptFileName = Brand<string, 'TransformedVueScriptFileName'>
