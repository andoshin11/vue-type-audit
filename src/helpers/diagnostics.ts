import _ts from 'typescript'
import { isTSVueFile, toRawVueFileName, isRawVueFile, toTSVueFIleName } from './path'
import { pos2location, createSourceFile, location2pos, getFullTextFromSnapshot } from './node'
import { SourcemapEntry, DiagnosticWithRange, FileEntry } from '../types'
import { readSystemFile } from './file'
import { findNode } from '../ast'
import { getOriginalPositionFor } from '../sourcemap'
import { Logger } from '../logger'

// TS diagnostic code
const REQUIRE_PROPS_CODE = 2554

const EXTERNAL_COMPONENT_IDENTIFIER_REGEXP = /^_component_.*$/

export const hasDiagRange = (diagnostic: _ts.Diagnostic): diagnostic is DiagnosticWithRange => {
  const { start, length, file } = diagnostic
  return typeof start === 'number' && typeof length === 'number' && !!file
}

export const translateTSVuefileDiagnostic = (diagnostic: _ts.Diagnostic, sourcemapEntry: SourcemapEntry, typeChecker: _ts.TypeChecker, ts: typeof _ts, logger?: Logger, fileEntry?: Map<string, { scriptSnapshot: _ts.IScriptSnapshot }>): _ts.Diagnostic => {
  const { file } = diagnostic

  if (!file || (!isTSVueFile(file.fileName) && !isRawVueFile(file.fileName))) {
    throw new Error('Invalid diagnostic format')
  }

  // Do nothing if the diagnostic has no range
  if (!hasDiagRange(diagnostic)) {
    return diagnostic
  }

  // Update sourcefile
  const tsVueFileName = isTSVueFile(file.fileName) ? file.fileName : toTSVueFIleName(file.fileName)
  const rawVueFileName = isRawVueFile(file.fileName) ? file.fileName : toRawVueFileName(file.fileName)
  let originalContent: string | undefined
  if (fileEntry) {
    const file = fileEntry.get(rawVueFileName)
    if (file) {
      originalContent = getFullTextFromSnapshot(file.scriptSnapshot)
    } else {
      originalContent = readSystemFile(rawVueFileName)
    }
  } else {
    originalContent = readSystemFile(rawVueFileName)
  }
  if (!originalContent) {
    throw new Error(`could not find file: ${rawVueFileName}`)
  }
  const sourcefile = createSourceFile(ts, originalContent, rawVueFileName)

  // Recover original position
  const { start, length } = diagnostic
  const content = file.getFullText()
  const startLC = pos2location(content, start)
  const endLC = pos2location(content, start + length)

  const originalStartLC = getOriginalPositionFor(tsVueFileName, startLC, sourcemapEntry)
  const originalEndLC = getOriginalPositionFor(tsVueFileName, endLC, sourcemapEntry)

  const originalStartPos = location2pos(originalContent, originalStartLC)
  const originalEndPos = location2pos(originalContent, originalEndLC)

  // Translate Message
  const translatedMessage = translateDiagnosticMessage(diagnostic, typeChecker, ts, logger)
  // console.log(ts.flattenDiagnosticMessageText(translatedMessage, '\n'))

  return {
    ...diagnostic,
    file: sourcefile,
    start: originalStartPos,
    length: originalEndPos - originalStartPos,
    messageText: translatedMessage
  }
}

export const translateDiagnosticMessage = (diagnostic: DiagnosticWithRange, typeChecker: _ts.TypeChecker, ts: typeof _ts, logger?: Logger): string | _ts.DiagnosticMessageChain => {
  const { code, start, messageText, file } = diagnostic
  if (logger) {
    logger.info(code)
    logger.info(messageText)
  }
  const translationTargetCodes = [REQUIRE_PROPS_CODE]
  let translatedMessage = messageText

  if (!translationTargetCodes.includes(code)) return translatedMessage

  if (diagnostic.code === REQUIRE_PROPS_CODE) {
    const node = findNode(file, start)
    if (!node) return translatedMessage
    const _createVNode = node.parent
    if (!ts.isCallExpression(_createVNode)) return translatedMessage
    const firstArg = _createVNode.arguments[0]
    if (!firstArg) return translatedMessage

    /**
     * Check if it's an external component node
     */

     // Check Node Type
    const typeInfo = typeChecker.getTypeAtLocation(firstArg)
    const isLiteralType = typeInfo.flags === ts.TypeFlags.StringLiteral
    if (!isLiteralType) return translatedMessage

    // Check external component Name
    const isExternalComponentNode = ts.isIdentifier(firstArg) && EXTERNAL_COMPONENT_IDENTIFIER_REGEXP.test(firstArg.escapedText.toString())
    if (!isExternalComponentNode) return translatedMessage

    const tagName = (typeInfo as _ts.LiteralType).value.toString()
    translatedMessage = `Missing required props for <${tagName}/> component.`

    // Check external component props type
    const requiredPropsNames = findRequiredPropNames(file, tagName, typeChecker, ts, logger)
    if (requiredPropsNames != null && requiredPropsNames.length >= 1) {
      translatedMessage = `Missing props [${requiredPropsNames.join(',')}] for <${tagName}/> component.`
    }

    return translatedMessage

  }

  return translatedMessage
}

export function findRequiredPropNames(file: _ts.SourceFile, componentName: string, typeChecker: _ts.TypeChecker, ts: typeof _ts, logger?: Logger): null | string[] {
  const findPropsDict = (node: _ts.Node): _ts.Node | undefined => {
    if (ts.isTypeAliasDeclaration(node) && node.name.escapedText.toString() === '___ExternalComponentsProps') {
      return node
    }
    return ts.forEachChild(node, findPropsDict)
  }
  const propsDict = findPropsDict(file)!

  const findTargetProperty = (node: _ts.Node): _ts.Node | undefined => {
    if (ts.isPropertySignature(node) && (node.name as _ts.Identifier).escapedText.toString() === componentName) {
      return node
    }
    return ts.forEachChild(node, findTargetProperty)
  }
  const targetComponent = findTargetProperty(propsDict)!
  const propType = typeChecker.getTypeAtLocation(targetComponent)

  const typeString = typeChecker.typeToString(propType)
  if (typeString === 'never') return null
  const requiredPropNames = typeString.split('|').map(s => s.trim())

  return requiredPropNames
}
