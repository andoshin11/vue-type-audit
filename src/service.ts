import _ts from 'typescript'
import * as fs from 'fs'
import * as path from 'path'
import { FileEntry, SourcemapEntry } from './types'
import { createHost, createService } from './languageService'
import { Reporter } from './reporter'
import { isTSVueFile, translateTSVuefileDiagnostic } from './helpers'
import { Logger } from './logger'

export class Service {
  private tsService: _ts.LanguageService
  private logger: Logger

  reporter: Reporter
  scriptVersions: FileEntry = new Map()
  sourcemapEntry: SourcemapEntry = new Map()

  constructor(public fileNames: string[], private compilerOptions: _ts.CompilerOptions, private ts: typeof _ts, debug: boolean = false) {
    const host = createHost(fileNames, compilerOptions, this.scriptVersions, this.sourcemapEntry, ts)
    this.tsService = createService(host, ts)
    this.reporter = new Reporter()
    this.logger = new Logger(debug ? 'info' : 'silent')
  }

  static fromConfigFile(configPath: string, ts: typeof _ts, debug: boolean = false): Service {
    const content = fs.readFileSync(configPath).toString();
    const parsed = ts.parseJsonConfigFileContent(
        JSON.parse(content),
        ts.sys,
        path.dirname(configPath)
    );
    const compilerOptions = parsed.options
    const fileNames = parsed.fileNames

    return new Service(fileNames, compilerOptions, ts, debug)
  }

  getSemanticDiagnostics() {
    const { ts, sourcemapEntry, tsService } = this
    const program = tsService.getProgram()
    if (!program) {
      throw new Error('could not create TS Program')
    }
    const typeChecker = program.getTypeChecker()

    const nativeSemanticDiagnostics = this.getNativeSemanticDiagnostics()

    // Tweak diagnostics for Vue files.
    const actualDiagnostics = nativeSemanticDiagnostics.map(d => {
      if (!d.file || !isTSVueFile(d.file.fileName)) return d

      return translateTSVuefileDiagnostic(d, sourcemapEntry, typeChecker, ts, this.logger)
    })

    return actualDiagnostics
  }

  getNativeSemanticDiagnostics() {
    const { tsService } = this
    const result = tsService.getProgram()!.getSemanticDiagnostics()
    return [...result]
  }
}
