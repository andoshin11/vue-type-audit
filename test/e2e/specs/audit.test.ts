import * as path from 'path'
import ts from 'typescript'
import { createServiceForTest } from '../service'
import { translateTSVuefileDiagnostic } from '../../../src/helpers/diagnostics'
import { isTSVueFile } from '../../../src/helpers'

function diagnosticSummary(diagnostic: ts.Diagnostic) {
  const { category, code, file, start, length, messageText } = diagnostic

  return {
    category,
    file: file ? file.fileName : null,
    code,
    start,
    length,
    messageText: ts.flattenDiagnosticMessageText(messageText, '\n')
  }
}

describe('audit', () => {
  test('should match snapshot', async () => {
    const localTS = require('../../../node_modules/typescript')
    const configPath = path.resolve(__dirname, '../project/tsconfig.json')
    const service = createServiceForTest(configPath)
    const diagnostics = service.getNativeSemanticDiagnostics()
    expect(diagnostics.map(diagnosticSummary)).toMatchSnapshot()
    const translated = diagnostics.map(d => translateTSVuefileDiagnostic(d, service.sourcemapEntry, service.tsService.getProgram()!.getTypeChecker(), localTS))
    expect(translated.map(diagnosticSummary)).toMatchSnapshot()
  })
})
