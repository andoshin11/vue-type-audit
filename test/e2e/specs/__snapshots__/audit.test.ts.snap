// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`audit should match snapshot 1`] = `
Array [
  Object {
    "category": 1,
    "code": 2339,
    "file": "/home/runner/work/vue-type-audit/vue-type-audit/test/e2e/project/src/App.vue.ts",
    "length": 6,
    "messageText": "Property 'length' does not exist on type 'number'.",
    "start": 414,
  },
  Object {
    "category": 1,
    "code": 2322,
    "file": "/home/runner/work/vue-type-audit/vue-type-audit/test/e2e/project/src/App.vue.ts",
    "length": 4,
    "messageText": "Type 'number' is not assignable to type 'string | undefined'.",
    "start": 2736,
  },
  Object {
    "category": 1,
    "code": 2339,
    "file": "/home/runner/work/vue-type-audit/vue-type-audit/test/e2e/project/src/App.vue.ts",
    "length": 6,
    "messageText": "Property 'length' does not exist on type 'number'.",
    "start": 2826,
  },
  Object {
    "category": 1,
    "code": 2554,
    "file": "/home/runner/work/vue-type-audit/vue-type-audit/test/e2e/project/src/App.vue.ts",
    "length": 35,
    "messageText": "Expected 2-6 arguments, but got 1.",
    "start": 2866,
  },
  Object {
    "category": 1,
    "code": 2322,
    "file": "/home/runner/work/vue-type-audit/vue-type-audit/test/e2e/project/src/App.vue.ts",
    "length": 8,
    "messageText": "Type '(val: string) => number' is not assignable to type '(val: number) => void'.
  Types of parameters 'val' and 'val' are incompatible.
    Type 'number' is not assignable to type 'string'.",
    "start": 3216,
  },
  Object {
    "category": 1,
    "code": 2345,
    "file": "/home/runner/work/vue-type-audit/vue-type-audit/test/e2e/project/src/App.vue.ts",
    "length": 28,
    "messageText": "Argument of type '{ onChanged: (val: number) => void; }' is not assignable to parameter of type 'VNodeProps & {} & { initialValue?: number | undefined; } & { onChange?: ((val: number) => void) | undefined; onInput?: ((...args: any) => void) | undefined; }'.
  Object literal may only specify known properties, but 'onChanged' does not exist in type 'VNodeProps & {} & { initialValue?: number | undefined; } & { onChange?: ((val: number) => void) | undefined; onInput?: ((...args: any) => void) | undefined; }'. Did you mean to write 'onChange'?",
    "start": 3391,
  },
  Object {
    "category": 1,
    "code": 2769,
    "file": "/home/runner/work/vue-type-audit/vue-type-audit/test/e2e/project/src/Counter.vue.ts",
    "length": 22,
    "messageText": "No overload matches this call.
  Overload 1 of 2, '(event: \\"input\\", ...args: any[]): void', gave the following error.
    Argument of type '\\"change\\"' is not assignable to parameter of type '\\"input\\"'.
  Overload 2 of 2, '(event: \\"change\\", val: number): void', gave the following error.
    Argument of type 'string' is not assignable to parameter of type 'number'.",
    "start": 433,
  },
]
`;

exports[`audit should match snapshot 2`] = `
Array [
  Object {
    "category": 1,
    "code": 2339,
    "file": "/home/runner/work/vue-type-audit/vue-type-audit/test/e2e/project/src/App.vue",
    "length": 6,
    "messageText": "Property 'length' does not exist on type 'number'.",
    "start": 907,
  },
  Object {
    "category": 1,
    "code": 2322,
    "file": "/home/runner/work/vue-type-audit/vue-type-audit/test/e2e/project/src/App.vue",
    "length": 4,
    "messageText": "Type 'number' is not assignable to type 'string | undefined'.",
    "start": 80,
  },
  Object {
    "category": 1,
    "code": 2339,
    "file": "/home/runner/work/vue-type-audit/vue-type-audit/test/e2e/project/src/App.vue",
    "length": 6,
    "messageText": "Property 'length' does not exist on type 'number'.",
    "start": 123,
  },
  Object {
    "category": 1,
    "code": 2554,
    "file": "/home/runner/work/vue-type-audit/vue-type-audit/test/e2e/project/src/App.vue",
    "length": 13,
    "messageText": "Missing props [\\"msg\\"] for <HelloWorld/> component.",
    "start": 149,
  },
  Object {
    "category": 1,
    "code": 2322,
    "file": "/home/runner/work/vue-type-audit/vue-type-audit/test/e2e/project/src/App.vue",
    "length": 6,
    "messageText": "Type '(val: string) => number' is not assignable to type '(val: number) => void'.
  Types of parameters 'val' and 'val' are incompatible.
    Type 'number' is not assignable to type 'string'.",
    "start": 306,
  },
  Object {
    "category": 1,
    "code": 2345,
    "file": "/home/runner/work/vue-type-audit/vue-type-audit/test/e2e/project/src/App.vue",
    "length": 21,
    "messageText": "Argument of type '{ onChanged: (val: number) => void; }' is not assignable to parameter of type 'VNodeProps & {} & { initialValue?: number | undefined; } & { onChange?: ((val: number) => void) | undefined; onInput?: ((...args: any) => void) | undefined; }'.
  Object literal may only specify known properties, but 'onChanged' does not exist in type 'VNodeProps & {} & { initialValue?: number | undefined; } & { onChange?: ((val: number) => void) | undefined; onInput?: ((...args: any) => void) | undefined; }'. Did you mean to write 'onChange'?",
    "start": 392,
  },
  Object {
    "category": 1,
    "code": 2769,
    "file": "/home/runner/work/vue-type-audit/vue-type-audit/test/e2e/project/src/Counter.vue",
    "length": 22,
    "messageText": "No overload matches this call.
  Overload 1 of 2, '(event: \\"input\\", ...args: any[]): void', gave the following error.
    Argument of type '\\"change\\"' is not assignable to parameter of type '\\"input\\"'.
  Overload 2 of 2, '(event: \\"change\\", val: number): void', gave the following error.
    Argument of type 'string' is not assignable to parameter of type 'number'.",
    "start": 580,
  },
]
`;
